### PR TITLE
Serialize/deserialize better with `System.Text.Json.JsonSerializer`

### PR DESCRIPTION
**Pull request by dahlia**
I implemented several [custom JSON converters] for types so that they are serialized better with [`System.Text.Json.JsonSerializer`].

[custom JSON converters]: https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to
[`System.Text.Json.JsonSerializer`]: https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer

### CODE REVIEW

**Review by tkiapril — DISMISSED** [review]

**Review by tkiapril — DISMISSED** [review]

**Review by moreal — DISMISSED** [review]

**Comment by dahlia** [comment]
@tkiapril @moreal I added two more types: `Currency` and `FungibleAssetValue`.  Could you take a look again?

File: CHANGES.md [hunk 1]
```diff
@@ -49,6 +49,10 @@ To be released.
 
      -  An `Address` became represented as a single hexadecimal string in JSON.
      -  A `BlockHash` became represented as a single hexadecimal string in JSON.
+     -  A `Currency` became represented as an object with values in JSON.
+        Note that it contains its `Hash` and it throws `JsonException`
+        if a JSON object to deserialize has an inconsitent `Hash` with
```
Comment (moreal) [inline CHANGES.md:54]: Maybe typo `inconsistent`

Comment (dahlia) [reply CHANGES.md]: Amended!

File: Libplanet.Tests/Assets/CurrencyTest.cs [hunk 2]
```diff
@@ -255,5 +255,39 @@ public void Serialize()
 
             Assert.Equal(bar, new Currency(bar.Serialize()));
         }
+
+        [Fact]
+        public void JsonSerialization()
+        {
+#pragma warning disable CS0618  // must test obsoleted Currency.Legacy() for backwards compatibility
+            var foo = Currency.Legacy("FOO", 2,  null);
```
Comment (moreal) [inline Libplanet.Tests/Assets/CurrencyTest.cs:263]: You can pass this part, but there was double spaces.

```suggestion
            var foo = Currency.Legacy("FOO", 2, null);
```

**Review by moreal — APPROVED** [review]

**Comment by moreal** [comment]
And the build seems broken with null-checker issues.

File: CHANGES.md [hunk 3]
```diff
@@ -43,14 +46,30 @@ To be released.
...
 ### Bug fixes
 
 ### Dependencies
 
- -  Upgraded *Bencodex* from [0.4.0][Bencodex 0.4.0] to [0.5.0][Bencodex 0.5.0].
-    [[#2283]]
- -  Upgraded *Bencodex* from [0.5.0][Bencodex 0.5.0] to [0.6.0][Bencodex 0.6.0].
-    [[#2298]]
+ -  Upgraded *Bencodex* from [0.4.0][Bencodex 0.4.0] to [0.6.0][Bencodex 0.6.0].
+    [[#2283]] [[#2298]]
```
Comment (moreal) [inline CHANGES.md:70]: I'm not sure about the CHANGELOG rule, Shouldn't that be like the below code?

```suggestion
    [[#2283], [#2298]]
```

**Review by moreal — APPROVED** [review]

File: Libplanet/Assets/Currency.cs [hunk 4]
```diff
@@ -311,7 +361,7 @@ private Currency(SerializationInfo info, StreamingContext context)
         private Currency(
```
Comment (tkiapril) [inline Libplanet/Assets/Currency.cs:361]: This private constructor was originally intended only for `Currency.Capped()` factory method, so if it is meant to also accompany `Currency.Uncapped()`, I guess you might have wanted to modify the `<summary />` for the constructor as well. 😅

**Review by tkiapril — APPROVED** [review]

File: CHANGES.md (hunk 3 above)
Comment (dahlia) [reply CHANGES.md:71]: As 0.42.0 was already released, I gave up this adjustment at all…

File: Libplanet/Assets/Currency.cs (hunk 4 above)
Comment (dahlia) [reply Libplanet/Assets/Currency.cs:362]: Adjusted the XML comment!

**Review by moreal — APPROVED** [review]

**Review by tkiapril — APPROVED** [review]

**Pull request merged**